sklearn/neighbors/base.py

```diff
@@ -212,7 +212,9 @@ def _fit(self, X):
             self._fit_method = 'kd_tree'
             return self
 
-        X = check_array(X, accept_sparse='csr')
+        nan_restricted = not callable(getattr(self, 'metric', None))
```

Comment:
this is not a very telling name. Maybe use ``allow_nan`` and pass ``not allow_nan`` to ``force_all_finite`` or call it ``forbid_nan``? Also, maybe add a comment on why we want that, like ``allow handling of NaN in user-defined metrics``" or something.